### PR TITLE
Fix for alluxio.logs.dir

### PR DESCRIPTION
**Pull request by flaming-archer**
### What changes are proposed in this pull request?
 After I append `"-Dalluxio.logs.dir"=<NEW_VALUE>"` to $ALLUXIO_JAVA_OPTS in alluxio-en.sh, found it is not work.
After set `export ALLUXIO_LOGS_DIR='<NEW_VALUE>'` ,it will work.
The reason is that in alluxio-config.sh,
````
ALLUXIO_LOGS_DIR="${ALLUXIO_LOGS_DIR:-${ALLUXIO_HOME}/logs}"
..
ALLUXIO_JAVA_OPTS+=" -Dalluxio.conf.dir=${ALLUXIO_CONF_DIR} -Dalluxio.logs.dir=${ALLUXIO_LOGS_DIR} -Dalluxio.user.logs.dir=${ALLUXIO_USER_LOGS_DIR}"

````
if you donnot set ALLUXIO_LOGS_DIR ,then the value of ALLUXIO_LOGS_DIR  is ${ALLUXIO_HOME}/logs and will overwrite alluxio.user.logs.dir.
So even if you definite  `-Dalluxio.logs.dir"=<NEW_VALUE>` in ALLUXIO_JAVA_OPTS , it will overwrite to default value and cannot work.

### Why are the changes needed?

Please clarify why the changes are needed. For instance,
  It actually not work.

### Does this PR introduce any user facing changes?

Please list the user-facing changes introduced by your change, including
no

### CODE REVIEW

**Comment by yuzhu** [comment]
@flaming-archer instead of changing the doc to indicate this does not work, is there a way to modify the alluxio-config.sh to make it work?

**Comment by flaming-archer** [comment]
> 

ok, i will modify.

**Comment by flaming-archer** [comment]
I fix 

> @flaming-archer instead of changing the doc to indicate this does not work, is there a way to modify the alluxio-config.sh to make it work?

I fixed, pls help to review.

File: libexec/alluxio-config.sh [hunk 1]
```diff
@@ -32,13 +32,23 @@ ALLUXIO_HOME=$(dirname $(dirname "${this}"))
...
-ALLUXIO_USER_LOGS_DIR="${ALLUXIO_USER_LOGS_DIR:-${ALLUXIO_LOGS_DIR}/user}"
 
 if [[ -e "${ALLUXIO_CONF_DIR}/alluxio-env.sh" ]]; then
   . "${ALLUXIO_CONF_DIR}/alluxio-env.sh"
 fi
 
+ALLUXIO_LOGS_DIR_JAVA_OPT_PRE="-Dalluxio.logs.dir="
+if [[  -n "${ALLUXIO_LOGS_DIR}" ]]; then
+  ALLUXIO_LOGS_DIR="${ALLUXIO_LOGS_DIR}"
+elif [[ ${ALLUXIO_JAVA_OPTS} == *${ALLUXIO_LOGS_DIR_JAVA_OPT_PRE}* ]]; then
```
Comment (yuzhu) [inline libexec/alluxio-config.sh:43]: could you add a comment what these three lines are doing?

Comment (flaming-archer) [reply libexec/alluxio-config.sh:46]: added.

Comment (flaming-archer) [reply libexec/alluxio-config.sh:46]: added

**Comment by yuzhu** [comment]
@jiacheliu3 seems like you had some experience with this issue before, ptal.

File: libexec/alluxio-config.sh [hunk 2]
```diff
@@ -32,13 +32,34 @@ ALLUXIO_HOME=$(dirname $(dirname "${this}"))
...
+  # If user set alluxio.logs.dir in ALLUXIO_JAVA_OPTS, use this.
+  # Extract "XXX -Dalluxio.logs.dir=<value> XXX".
+  # First get "<value> XXX".
+  ALLUXIO_LOGS_DIR_TMP=${ALLUXIO_JAVA_OPTS#*${ALLUXIO_LOGS_DIR_JAVA_OPT_PRE}}
+  # Then get "<value>".
+  ALLUXIO_LOGS_DIR=$(echo "${ALLUXIO_LOGS_DIR_TMP}"|awk '{print $1}')
+  # Set ALLUXIO_LOGS_DIR_FROM_ALLUXIO_JAVA_OPTS_TAG to true is let ALLUXIO_JAVA_OPTS
+  # not append -Dalluxio.user.logs.dir=<value> later, since it already have.
+  ALLUXIO_LOGS_DIR_FROM_ALLUXIO_JAVA_OPTS_TAG="true"
```
Comment (jiacheliu3) [inline libexec/alluxio-config.sh:55]: I haven't give it too much thought, but I believe the logic doesn't have to be this complicated?

The issue is the same for all environment var translations. For example if you define `ALLUXIO_WORKER_HOSTNAME=a` but `-Dalluxio.worker.hostname=b`, we have the line `ALLUXIO_WORKER_JAVA_OPTS+=-Dalluxio.worker.hostname=${ALLUXIO_WORKER_HOSTNAME}` so one overrides the other. For alluxio, it often goes without saying that you use the env var and rely on them being translated to a java property somewhere.

I think one easy way out is we ask the users to use `ALLUXIO_LOGS_DIR` instead of `-Dalluxio.logs.dir`.

Otherwise we can search if `alluxio.logs.dir` is set in java properties already so we don't set later.

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Thanks for the contribution!

Adding @Xenorith for discussion because the problem is actually more general than just this property.

File: libexec/alluxio-config.sh (hunk 2 above)
Comment (flaming-archer) [reply libexec/alluxio-config.sh:55]: I think my modify logic is the same as 

```
I think one easy way out is we ask the users to use ALLUXIO_LOGS_DIR instead of -Dalluxio.logs.dir.

Otherwise we can search if alluxio.logs.dir is set in java properties already so we don't set later.
```
But we must set the value  of `alluxio.logs.dir` from  java property.
Because many shell scrips relay on this var such as `alluxio-start.sh`, `alluxio-master.sh` etc.
There are snippets in alluxio-start.sh
```
ensure_dirs() {
  if [[ ! -d "${ALLUXIO_LOGS_DIR}" ]]; then
    echo "ALLUXIO_LOGS_DIR: ${ALLUXIO_LOGS_DIR}"
    mkdir -p ${ALLUXIO_LOGS_DIR}
  fi
}
```
and 
```
    echo "Starting job master @ $(hostname -f). Logging to ${ALLUXIO_LOGS_DIR}"

```

So if we do not define a env var but define in java property , and I change the code to use java property actually , then the script will be all wrong.

Comment (Xenorith) [reply libexec/alluxio-config.sh:55]: i do not think it is a good idea to use the env var value of ALLUXIO_LOGS_DIR in later shell scripts. the env var is not the source of truth. can we run `bin/alluxio getConf alluxio.logs.dir` to get this value when it is time to run the `mkdir -p` command?

**Review by Xenorith — COMMENTED** [review]
i spent an hour looking and thinking about this problem and there are multiple areas of improvement. for this particular problem, i think the simplest way to approach it is:
- i suggest to only use the value of `ALLUXIO_LOGS_DIR` env var to set `JAVA_OPTS` and not use it for later shell scripts. instead, it would be better to ask alluxio for its configured value by running `bin/alluxio getConf` since this ensures the value being used is consistent
- alluxio-config.sh mostly follows the pattern of "if an env var is set, append the corresponding property in the process' java opts". the exception is in line 70 where it always sets `alluxio.conf.dir`, `alluxio.logs.dir`, and `alluxio.user.logs.dir`. this should also be made optional
  - if ALLUXIO_LOGS_DIR is set by the user, then append the properties for `alluxio.logs.dir` and `alluxio.user.logs.dir`
  - ALLUXIO_CONF_DIR is a special case because there must be a value for it in order to source alluxio-env.sh in line 39. so if the user did not provide it, then we must default to `ALLUXIO_HOME/conf`. this means the user should never set `-Dalluxio.conf.dir` as a java opt

i also see other inconsistencies such as the case of using ALLUXIO_MASTER_JAVA_OPTS.  if you set `-Dalluxio.logs.dir` in ALLUXIO_MASTER_JAVA_OPTS, it will be respected by the alluxio master process but alluxio-start.sh will use the wrong value. so the correct order of operations should actually be like:
- start with default java opts for a process
- append shared java opts = ALLUXIO_JAVA_OPTS
- append process specific java opts, ex. ALLUXIO_MASTER_JAVA_OPTS
- append env vars such that they always take precedence

this needs a more thorough review in addition to looking at how the logs property is behaving

**Comment by apc999** [comment]
some context on this @flaming-archer @yuzhu @Xenorith 

- For most of the Alluxio properties (included in https://docs.alluxio.io/os/user/stable/en/reference/Properties-List.html), they are only needed in Java land, the corresponding env vars (e.g., `ALLUXIO_MASTER_HOSTNAME` which corresponds to `alluxio.master.hostname`) only exists to facilitate controlling the setting in `alluxio-env.sh`, which is typically useful for deployment tools. So the recommended place to host property assignment is in `alluxio-site.properties` file, and this file is only loaded in Java land after the Java process is launched --- they are not known by the shells at JVM starting time

- However, `alluxio.log.dir` (and its derivatives like `Dalluxio.user.logs.dir`) is different. They are required to be known as Java opts to pass to log4j library --- we previously tried to set these properties like other normal ones after JVM started, which is not working.

In short, `alluxio.log.dir` is a special property similar to `alluxio.conf.dir`, unlike other properties.

If we really want to fix this issue, a hacky but simple solution is to change
https://github.com/Alluxio/alluxio/blob/b7b239658dd3daf0fc998835dd78d39de3f22b1b/libexec/alluxio-config.sh#L70

to 

```
 ALLUXIO_JAVA_OPTS=" -Dalluxio.conf.dir=${ALLUXIO_CONF_DIR} -Dalluxio.logs.dir=${ALLUXIO_LOGS_DIR} -Dalluxio.user.logs.dir=${ALLUXIO_USER_LOGS_DIR} ${ALLUXIO_JAVA_OPTS}" 
```

as for most JVM handles conflicted java opt def by having the later definition win

**Comment by Xenorith** [comment]
i see, so `-Dalluxio.logs.dir` must be set in the java opts, even if it is not provided as part of the env var or as part of ALLUXIO_JAVA_OPTS. prepending the default value to `ALLUXIO_JAVA_OPTS` satisfies this.

this still runs into the problem of needing to create the logs directory based on the value of ALLUXIO_LOGS_DIR. if `ALLUXIO_LOGS_DIR` is not defined and `-Dalluxio.logs.dir` is set in `ALLUXIO_JAVA_OPTS`, then alluxio-start.sh will incorrectly run `mkdir` in `$ALLUXIO_HOME/logs`, so the workaround to run `getConf` is still needed to know what the true value of the logs dir is

**Comment by apc999** [comment]
> this still runs into the problem of needing to create the logs directory based on the value of ALLUXIO_LOGS_DIR. if `ALLUXIO_LOGS_DIR` is not defined and `-Dalluxio.logs.dir` is set in `ALLUXIO_JAVA_OPTS`, then alluxio-start.sh will incorrectly run `mkdir` in `$ALLUXIO_HOME/logs`, so the workaround to run `getConf` is still needed to know what the true value of the logs dir is

In `alluxio-start.sh`, there is no problemto call `alluxio getConf` to retrieve the resolved value of this property. 
I have no problem with that. Leaving this to you guys

**Comment by flaming-archer** [comment]
Hello all, in my env, whether set alluxio.log.dir to java property or not, type 
```
bin/alluxio getConf alluxio.log.dir
```
always display
```
alluxio.log.dir is not a valid configuration key
```
And I think  if we use this
```
 ALLUXIO_JAVA_OPTS=" -Dalluxio.conf.dir=${ALLUXIO_CONF_DIR} -Dalluxio.logs.dir=${ALLUXIO_LOGS_DIR} -Dalluxio.user.logs.dir=${ALLUXIO_USER_LOGS_DIR} ${ALLUXIO_JAVA_OPTS}" 
```
There is a problem that the user use both the jvm var and env var ,before the env is effective but after the jvm will be effective. So when they upgrade the alluxio  since merge this code, they will find the logpath  is changed.

Since it is so diffcult, can we back to my first commit  https://github.com/Alluxio/alluxio/pull/14837/commits/f34c2bb71eca3c018e6712ece54326df82eb6e61.

We do not support set jvm property to controlle.

**Comment by flaming-archer** [comment]
> i see, so `-Dalluxio.logs.dir` must be set in the java opts, even if it is not provided as part of the env var or as part of ALLUXIO_JAVA_OPTS. prepending the default value to `ALLUXIO_JAVA_OPTS` satisfies this.
> 
> this still runs into the problem of needing to create the logs directory based on the value of ALLUXIO_LOGS_DIR. if `ALLUXIO_LOGS_DIR` is not defined and `-Dalluxio.logs.dir` is set in `ALLUXIO_JAVA_OPTS`, then alluxio-start.sh will incorrectly run `mkdir` in `$ALLUXIO_HOME/logs`, so the workaround to run `getConf` is still needed to know what the true value of the logs dir is

I agree with ` then alluxio-start.sh will incorrectly run mkdir in $ALLUXIO_HOME/logs`.
I test and  logpath is the default localtion. And the console display
```
$ sh bin/alluxio-start.sh -a master
Starting master @ xxxxxx Logging to /opt/alluxio/logs
```

**Comment by Xenorith** [comment]
> Hello all, in my env, whether set alluxio.log.dir to java property or not, type
> 
> ```
> bin/alluxio getConf alluxio.log.dir
> ```
> 
> always display
> 
> ```
> alluxio.log.dir is not a valid configuration key
> ```

this is expected because there is a typo in your command. it should be `alluxio.logs.dir` with a `s` after `log`. this command worked when i inserted it in `alluxio-start.sh`, could you try again?

> 
> And I think if we use this
> 
> ```
>  ALLUXIO_JAVA_OPTS=" -Dalluxio.conf.dir=${ALLUXIO_CONF_DIR} -Dalluxio.logs.dir=${ALLUXIO_LOGS_DIR} -Dalluxio.user.logs.dir=${ALLUXIO_USER_LOGS_DIR} ${ALLUXIO_JAVA_OPTS}" 
> ```
> 
> There is a problem that the user use both the jvm var and env var ,before the env is effective but after the jvm will be effective when they upgrade the alluxio since merge this code. They will find the logpath is changed.
> 

yes there is a backcompat break in this hopefully rare scenario, where there is a user that set both the jvm option and the env var as well as using different values for each. however, this is a bug and i prefer that we do the correct thing rather than keep the existing buggy behavior. it can lead to more inconsistencies and confusion later; another user may raise a future issue "why does config property x behave one way but the logs config property behave the opposite way?" @apc999 @jiacheliu3 do you have any opinions on this?

> Since it is so diffcult, can we back to my first commit [f34c2bb](https://github.com/Alluxio/alluxio/commit/f34c2bb71eca3c018e6712ece54326df82eb6e61).
> 
> We do not support set jvm property to controlle.

i believe it is possible to support without too much additional work, so i would still encourage that we do the right thing for the long term health of the usage/behavior. i would like the final behavior to be described as "we encourage users to use the dedicated env var which takes precedence, but it is also possible to set this configuration through the java opts env var if the dedicated env var is not set". this order of configuration precedence is more intuitive than "it looks like there are multiple ways to set a configuration, but one will not work and you won't get a warning or error about it"

**Comment by jiacheliu3** [comment]
I believe the env var should take precedence over java property during the startup.

In other words, the behavior below is undesirable and unintended(basically a bug), so we should remove that. Is your environment currently set up that way? If there are two ways to set a property, setting the two ways to different values does not make much sense to me. You will have to remove one or make them match sooner or later.
> There is a problem that the user use both the jvm var and env var ,before the env is effective but after the jvm will be effective when they upgrade the alluxio since merge this code. They will find the logpath is changed.

+1 for correcting the behavior and removing the current buggy behavior. To compensate who are relying on the jvm var, is it possible to output a warning that in the scripts only the ALLUXIO_LOGS_DIR should be used?

**Comment by Xenorith** [comment]
this is one of those "how nice should we be to the user" and i'm inclined to say that we don't need to write extra regex logic to parse the contents of the java opt value for the sake of printing a warning. the jvm arg is still expected to work as long as the env var is not set; it's still valid but takes lower precedence, similar to how a credentials chain would work.

**Comment by flaming-archer** [comment]
> this is one of those "how nice should we be to the user" and i'm inclined to say that we don't need to write extra regex logic to parse the contents of the java opt value for the sake of printing a warning. the jvm arg is still expected to work as long as the env var is not set; it's still valid but takes lower precedence, similar to how a credentials chain would work.

I think if we want to get the value from jvm opt,  are they any other ways to extract besides  `extra regex logic`? 
If user donnot set env var but set jvm opt. so the work to extract value must to do. And must set the value to ALLUXIO_LOGS_DIR.

**Comment by apc999** [comment]
close this PR due to my latest change https://github.com/Alluxio/alluxio/pull/14980
